feat: nicer skip context for macro/attribute

tommilligan opened the pull request:
Follow up from https://github.com/rust-lang/rustfmt/pull/5347#discussion_r895000579 in discussion with @ytmimi 

Refactors the way named items are skipped such that:

- skipping each item is constant time
- implementation is the same for attributes/macros
- either all items can be skipped, or a set of named items

This should result in no change in behaviour.

ytmimi commented on src/skip.rs:
```diff
@@ -2,41 +2,60 @@
+
+impl SkipNameContext {
+    pub(crate) fn append(&mut self, values: Vec<String>) {
+        self.values.extend(values);
```
once `self.all=true` does it make sense to extend? Could this be:

```suggestion
        if !self.all {
            self.values.extend(values);
        }
```

tommilligan replied:
Agreed - I refactored the inner types, but made sure the same logic applies.

ytmimi commented on src/skip.rs:
```diff
@@ -2,41 +2,60 @@
-        self.macros.extend(other.into_iter());
+    pub(crate) fn update(&mut self, other: Self) {
+        self.all = self.all || other.all;
+        self.values.extend(other.values);
```
Since the extend logic is implemented in `append` could we change this to
```suggestion
        self.append(other.values);
```

ytmimi commented on src/visitor.rs:
```diff
@@ -775,10 +775,10 @@ impl<'b, 'a: 'b> FmtVisitor<'a> {
         for macro_selector in config.skip_macro_invocations().0 {
             match macro_selector {
                 MacroSelector::Name(name) => macro_names.push(name.to_string()),
-                MacroSelector::All => skip_context.all_macros = true,
+                MacroSelector::All => skip_context.macros.set_all(true),
             }
         }
-        skip_context.update_macros(macro_names);
+        skip_context.macros.append(macro_names);
```
Along with the comments above, I wonder if it makes sense to avoid calling `to_string` when "*" is in the `skip_macro_invocations` list by making the following update: 

```rust
let contains_all = config.skip_macro_invocations().0.any(MacroSelector::is_all);

if contains_all {
    skip_context.macros.set_all(true);
} else {
    skip_context.macros.append(config.skip_macro_invocations().0.collect());
}
```

The code might need to be written a little differently, but I hope the general idea is understood.

tommilligan replied:
I think realistically it's unlikely we see a case that makes it worth this optimization, at the cost of reading the vec twice. My expectation is that `"*"`, if given, will be the only string in the vec, as it makes all other names redundant. The unit tests that check combinations of the two are just for completeness.

tommilligan commented:
Updated to use the `Extend` trait and rename a bunch of `append` calls to `extend` for correctness.

Refactored the inner type to an enum, since it matches the behaviour nicely.

ytmimi commented:
@tommilligan did you push the changes? I don't think I"m seeing anything different than last time. I'm also not seeing any impls for `Extend`. Am I missing something?

tommilligan commented:
> @tommilligan did you push the changes? I don't think I"m seeing anything different than last time. I'm also not seeing any impls for `Extend`. Am I missing something?

Sorry - now pushed. Was working on the train, looks like the connection failed and I didn't notice, sorry.

calebcartwright commented on src/visitor.rs:
```diff
@@ -775,10 +775,10 @@ impl<'b, 'a: 'b> FmtVisitor<'a> {
             match macro_selector {
                 MacroSelector::Name(name) => macro_names.push(name.to_string()),
-                MacroSelector::All => skip_context.all_macros = true,
+                MacroSelector::All => skip_context.macros.skip_all(),
```
I'd think we could technically break after this arm because if we reach it then it's a bit moot whether there's additional entries. However, it shouldn't really matter. I don't even a super villain could be pathological enough to make it matter

```suggestion
                MacroSelector::All => {
                    skip_context.macros.skip_all();
                    break;
                }
```

calebcartwright reviewed (approved):
This lgtm but will hold off on merging in case you want to take a final pass @ytmimi

ytmimi reviewed (approved):
@tommilligan thanks for your work on this! I like the update and I'm ready to get this pulled in!
